awx/main/wsrelay.py

```diff
@@ -0,0 +1,307 @@
+        raise RuntimeError("Implement me")
+
+    async def connect(self):
+        from awx.main.consumers import WebsocketSecretAuthHelper  # noqa
+
+        logger.debug(f"Connection attempt from {self.name} to {self.remote_host}")
+
+        '''
+        Can not put get_channel_layer() in the init code because it is in the init
+        path of channel layers i.e. RedisChannelLayer() calls our init code.
+        '''
+        if not self.channel_layer:
+            self.channel_layer = get_channel_layer()
+
+        uri = f"{self.protocol}://{self.remote_host}:{self.remote_port}/websocket/relay/"
+        timeout = aiohttp.ClientTimeout(total=10)
+
+        secret_val = WebsocketSecretAuthHelper.construct_secret()
+        try:
+            async with aiohttp.ClientSession(headers={'secret': secret_val}, timeout=timeout) as session:
+                async with session.ws_connect(uri, ssl=self.verify_ssl, heartbeat=20) as websocket:
+                    logger.info(f"Connection from {self.name} to {self.remote_host} established.")
+                    self.stats.record_connection_established()
+                    self.connected = True
+                    await self.run_connection(websocket)
+        except asyncio.CancelledError:
+            # TODO: Check if connected and disconnect
+            # Possibly use run_until_complete() if disconnect is async
+            logger.warning(f"Connection from {self.name} to {self.remote_host} cancelled.")
+        except client_exceptions.ClientConnectorError as e:
+            logger.warning(f"Connection from {self.name} to {self.remote_host} failed: '{e}'.", exc_info=True)
+        except asyncio.TimeoutError:
+            logger.warning(f"Connection from {self.name} to {self.remote_host} timed out.")
+        except Exception as e:
+            # Early on, this is our canary. I'm not sure what exceptions we can really encounter.
+            logger.warning(f"Connection from {self.name} to {self.remote_host} failed for unknown reason: '{e}'.", exc_info=True)
+        else:
+            logger.info(f"Connection from {self.name} to {self.remote_host} lost, but no exception was raised.")
+        finally:
+            self.connected = False
+            self.stats.record_connection_lost()
+
+    def start(self):
+        self.async_task = self.event_loop.create_task(self.connect())
+        return self.async_task
+
+    def cancel(self):
+        self.async_task.cancel()
+
+    async def run_connection(self, websocket: aiohttp.ClientWebSocketResponse):
+        # create a dedicated subsystem metric producer to handle local subsystem
+        # metrics messages
+        # the "metrics" group is not subscribed to in the typical fashion, so we
+        # just explicitly create it
+        producer = self.event_loop.create_task(self.run_producer("metrics", websocket, "metrics"))
+        self.producers["metrics"] = {"task": producer, "subscriptions": {"metrics"}}
```

Comment:
@fosterseth I might be missing something here but is this kind of looks like we are only running one metrics producer and then overwriting it every time we establish a new connection. Look below where we create the "normal" producer, we scope it to the name of the remote connection:

```
name = f"{self.remote_host}-{group}"
```

I think what you want might want this to be something like:
```suggestion
        self.producers[f"{self.remote_host}-metrics"] = {"task": producer, "subscriptions": {"metrics"}}
```